hoomd/md/force.py

```diff
@@ -117,212 +118,144 @@ def additional_virial(self):
-                self.fvec[1],
-                self.fvec[2],
-                self.tvec[0],
-                self.tvec[1],
-                self.tvec[2],
-            )
-        elif tag is not None:
-            self.cppForce.setParticleForce(
-                tag,
-                self.fvec[0],
-                self.fvec[1],
-                self.fvec[2],
-                self.tvec[0],
-                self.tvec[1],
-                self.tvec[2],
-            )
-        else:
-            self.cppForce.setForce(
-                self.fvec[0],
-                self.fvec[1],
-                self.fvec[2],
-                self.tvec[0],
-                self.tvec[1],
-                self.tvec[2],
-            )
+    Note:
+        All force data buffers are MPI rank local, so in simulations with MPI,
+        only the data for a single rank is available.
 
-    R""" Set a python callback to be called before the force is evaluated
+    Note:
+        Access to the force buffers is constant (O(1)) time.
+    """
 
-    Args:
-        callback (`callable`) The callback function
+    def __init__(self):
+        self._in_context_manager = False
+        self._state = None  # to be set on attaching
 
-     Examples:
-        const = force.constant(fx=0.4, fy=1.0, fz=0.5)
+    def _attach(self):
+        self._state = self._simulation.state
+        self._cpp_obj = _md.CustomForceCompute(self._state._cpp_sys_def)
+        self._cpp_obj.setCallback(self.set_forces)
+        super()._attach()
 
-        def updateForces(timestep):
-            global const
-            const.setForce(tag=1, fvec=(1.0*timestep,2.0*timestep,3.0*timestep))
+    @property
+    def cpu_local_force_arrays(self):
+        """hoomd.data.LocalForceAccess: Expose force arrays on the CPU.
 
-        const.set_callback(updateForces)
-        run(100)
+        Provides direct access to the force, potential energy, torque, and
+        virial data of the particles in the system on the cpu through a context
+        manager. All data is MPI rank-local.
 
-        # Reset the callback
-        const.set_callback(None)
-    """
+        The `hoomd.data.LocalForceAccess` object returned by this property has
+        four arrays through which one can modify the force data:
+
+        Examples::
 
-    def set_callback(self, callback=None):  # noqa - will be ported to v3
-        self.cppForce.setCallback(callback)
+            with self.cpu_local_force_arrays as arrays:
+                arrays.force = ...
+                arrays.potential_energy = ...
+                arrays.torque = ...
+                arrays.virial = ...
+        """
+        if self._in_context_manager:
+            raise RuntimeError("Cannot enter cpu_local_force_arrays context "
+                               "manager inside another local_force_arrays "
+                               "context manager")
+        return hoomd.data.ForceLocalAccess(self)
+
+    @property
+    def gpu_local_force_arrays(self):
```

Comment:
Just for my own clarification, although it might be beneficial to users, with the internals of HOOMD, integrators won't touch the GPU so there's no need to set this unless you're doing some other computes with a GPU implementation correct?

Reply:
I'm not quite sure what you mean by "integrators won't touch the GPU"; the integration methods have GPU implementations in MD. This property isn't intended to be set. If your simulation uses other computes who need access to the data buffers we expose here, that will still work correctly because the ArrayHandles on the force data buffers are released as soon as the context manager exits.